File: solutions/python/binary-search.py
# id: 008
# title: Binary Search
# approach: Iterative Binary Search
# time: O(log n) | space: O(1)
# author: chron303

import sys

def main():
    data = sys.stdin.read().split()
    n = int(data[0])
    nums = list(map(int, data[1:n+1]))
    target = int(data[n+1])

    left, right = 0, n - 1
    while left <= right:
        mid = left + (right - left) // 2
        if nums[mid] == target:
            print(mid)
            return
        elif nums[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    print(-1)

if __name__ == "__main__":
    main()

File: solutions/cpp/binary-search.cpp
// id: 008
// title: Binary Search
// approach: Iterative Binary Search
// time: O(log n) | space: O(1)
// author: chron303

#include <bits/stdc++.h>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    int target;
    cin >> target;

    int left = 0, right = n - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (nums[mid] == target) {
            cout << mid << endl;
            return 0;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    cout << -1 << endl;
    return 0;
}
